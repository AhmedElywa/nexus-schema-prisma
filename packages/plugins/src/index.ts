export * from './select';
export * from './sdlInputs';
