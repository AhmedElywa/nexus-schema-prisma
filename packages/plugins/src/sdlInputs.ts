import { GraphQLSchema } from 'graphql';
import { writeFileSync } from 'fs';

export const generateGraphQlSDLFile = (schema: GraphQLSchema, path = 'schema.graphql') => {
  const { printSchema } = require('graphql');
  writeFileSync(path, printSchema(schema));
};
