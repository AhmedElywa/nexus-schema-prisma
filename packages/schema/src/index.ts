export * from '@paljs/types';
export * from './json';
export * from './camelCase';
export * from './typescript';
export * from './PrismaReader';
