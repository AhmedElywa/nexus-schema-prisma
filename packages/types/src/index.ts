export * from './cli';
export * from './generatedSchema';
export * from './generator';
export * from './configFile';
export * from './adminSchema';
export * from './dmmf';
