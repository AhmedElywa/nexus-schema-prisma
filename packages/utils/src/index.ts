export * from './dmmf';
export * from './schemaInputs';
export { getConfig, getEnvPaths, tryLoadEnvs, getDMMF } from '@prisma/internals';
