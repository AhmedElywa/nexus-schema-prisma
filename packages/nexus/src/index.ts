import { PrismaSelect } from '@paljs/plugins';
import type { DMMF } from '@paljs/types';
import { objectType, plugin } from 'nexus';
import { NexusAcceptedTypeDef } from 'nexus/dist/builder';
import { adminNexusSchemaSettings } from './admin';
import type { Settings } from './settings';
import { getScalars } from './defaultScalar';
import { AllNexusOutputTypeDefs } from 'nexus/dist/definitions/wrapping';

export { Settings };

export const paljs = (settings?: Settings) =>
  plugin({
    name: 'paljs',
    description:
      'paljs plugin to add Prisma select to your resolver and prisma admin queries and mutations and all models input types',
    onInstall(builder) {
      const nexusSchemaInputs: (NexusAcceptedTypeDef | AllNexusOutputTypeDefs)[] = [
        objectType({
          name: 'BatchPayload',
          definition(t) {
            t.nonNull.int('count');
          },
        }),
        ...getScalars(settings?.excludeScalar),
      ];

      if (settings?.includeAdmin) {
        nexusSchemaInputs.push(...adminNexusSchemaSettings(settings?.adminSchemaPath));
      }
      for (const type of nexusSchemaInputs) {
        builder.addType(type);
      }
    },
    onCreateFieldResolver() {
      return async (root, args, ctx, info: any, next) => {
        ctx.select = new PrismaSelect(info, {
          ...settings?.prismaSelectOptions,
          dmmf: settings?.prismaSelectOptions?.dmmf as Omit<DMMF.Document, 'schema'>[],
        });
        return await next(root, args, ctx, info);
      };
    },
  });
