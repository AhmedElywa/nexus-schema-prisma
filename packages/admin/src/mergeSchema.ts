import { Schema, SchemaModel, SchemaField } from './types';
import { SchemaObject, Model, Field } from '@prisma-tools/schema';

export function mergeSchema(object: SchemaObject, schema: Schema): Schema {
  const newSchema: Schema = {
    models: [],
    enums: object.enums,
  };
  object.models.forEach((item) => {
    const schemaItem = schema.models.find((model) => model.id === item.name);
    if (!schemaItem) {
      if (checkIdFieldExist(item)) {
        newSchema.models.push(handleNewModel(item));
      }
    } else {
      const newItem: SchemaModel = {
        ...schemaItem,
        idField:
          schemaItem.idField ?? item.fields.find((field) => field.isId)?.name,
        displayFields: schemaItem.displayFields ?? [
          item.fields.find((field) => field.isId)?.name,
        ],
        fields: [],
      };
      item.fields.forEach((field) => {
        const schemaField = schemaItem.fields.find(
          (item) => item.name === field.name,
        );
        if (!schemaField) {
          newItem.fields.push(handleNewField(field, schemaItem.name));
        } else {
          newItem.fields.push({
            ...schemaField,
            ...getOriginalField(field, schemaItem.name),
          });
        }
      });
      newItem.fields.sort((a, b) => a.order - b.order);
      newSchema.models.push(newItem);
    }
  });
  return newSchema;
}

function checkIdFieldExist(model: Model) {
  return !!model.fields.find((field) => field.isId);
}

function handleNewModel(model: Model) {
  const newItem: SchemaModel = {
    id: model.name,
    name: getTitle(model.name),
    idField: model.fields.find((field) => field.isId)!.name,
    displayFields: [model.fields.find((field) => field.isId)!.name],
    create: true,
    update: true,
    delete: true,
    fields: [],
  };
  model.fields.forEach((field) => {
    newItem.fields.push(handleNewField(field, model.name));
  });
  return newItem;
}

const defaultField = ['id', 'createdAt', 'updatedAt'];

function handleNewField(field: Field, modelName: string): SchemaField {
  return {
    ...getOriginalField(field, modelName),
    title: getTitle(field.name),
    create: !defaultField.includes(field.name) && !field.relationField,
    update: !defaultField.includes(field.name) && !field.relationField,
    read: true,
    filter: true,
    sort: true,
    order: -1,
  };
}

function getTitle(id: string) {
  const split = id.split(/(?=[A-Z])/);
  split[0] = split[0].charAt(0).toUpperCase() + split[0].slice(1);
  return split.join(' ');
}

function getOriginalField(
  field: Field,
  modelName: string,
): Omit<Field, 'relation'> & { id: string } {
  delete field.relation;
  return {
    id: modelName + '.' + field.name,
    ...field,
  };
}
