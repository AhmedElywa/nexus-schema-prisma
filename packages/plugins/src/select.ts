import { GraphQLResolveInfo } from 'graphql';
import { DMMF } from '@paljs/types';
import { parseResolveInfo } from 'graphql-parse-resolve-info';

/**
 * Convert `info` to select object accepted by `prisma client`.
 * @param info - GraphQLResolveInfo.
 * @example
 * // Graphql query
 * {
 *    findManyUser{
 *      id
 *      posts(where: { title: { contains: "a" } }, first: 10) {
 *        id
 *        comments{
 *          id
 *        }
 *      }
 *    }
 * }
 * // convert to
 * {
 *  select: {
 *    id: true,
 *    posts: {
 *      select: { id: true, comments: { select: { id: true } } },
 *      where: { title: { contains: "a" } },
 *      first: 10
 *    }
 *  }
 * }
 *
 * // Use
 *
 * const select = new PrismaSelect(info);
 *
 * prisma.user.findMany({
 *  ...args,
 *  ...select.value,
 * })
 *
 **/
export class PrismaSelect {
  private availableArgs = ['where', 'orderBy', 'skip', 'cursor', 'take', 'distinct'];
  private allowedProps = ['_count'];
  private isAggregate = false;

  constructor(
    private info: GraphQLResolveInfo,
    private options?: {
      /*
       * you can pass object with your models and what the fields you need to include for every model even if user not requested in GraphQL query.
       * @example
       * const defaultFields = {
       *    User: { id: true, name: true },
       *    Type: { id: true, descriptionRaw: true },
       *    Post: { id: true, body: true },
       *    // as function you can check if client select some fields to add another to default fields
       *    Account: (select) => select.name ? {firstname: true, lastname: true} : {}
       * }
       * */
      defaultFields?: {
        [key: string]: { [key: string]: boolean } | ((select: any) => { [key: string]: boolean });
      };
      /*
       * array of dmmf object import from generated prisma client default
       * @example
       * import {Prisma} from './customPath';
       * import {Prisma as Prisma2} from './customPath2';
       * {
       *  dmmf: [Prisma.dmmf, Prisma2.dmmf]
       * }
       * */
      dmmf?: Omit<DMMF.Document, 'schema'>[];
    },
  ) {}

  get value() {
    const returnType = this.info.returnType.toString().replace(/]/g, '').replace(/\[/g, '').replace(/!/g, '');
    this.isAggregate = returnType.includes('Aggregate');
    return this.valueWithFilter(returnType);
  }

  get dataModel() {
    const models: DMMF.Model[] = [];
    if (this.options?.dmmf) {
      this.options?.dmmf.forEach((doc) => {
        models.push(...doc.datamodel.models);
      });
    } else {
      const { Prisma } = require('@prisma/client');
      if (Prisma.dmmf && Prisma.dmmf.datamodel) {
        const models: DMMF.Model[] = Prisma.dmmf.datamodel.models;
        models.push(...models);
      }
    }
    return models;
  }

  get defaultFields() {
    return this.options?.defaultFields;
  }

  private get fields() {
    return parseResolveInfo(this.info);
  }

  private static getModelMap(docs?: string, name?: string) {
    const value = docs?.match(/@PrismaSelect.map\(\[(.*?)\]\)/);
    if (value && name) {
      const asArray = value[1]
        .replace(/ /g, '')
        .split(',')
        .filter((v) => v);
      return asArray.includes(name);
    }
    return false;
  }

  private model(name?: string) {
    return this.dataModel.find((item) => item.name === name || PrismaSelect.getModelMap(item.documentation, name));
  }

  private field(name: string, model?: DMMF.Model) {
    return model?.fields.find((item) => item.name === name);
  }

  static isObject(item: any) {
    return item && typeof item === 'object' && !Array.isArray(item);
  }

  static mergeDeep(target: any, ...sources: any[]): any {
    if (!sources.length) return target;
    const source: any = sources.shift();

    if (PrismaSelect.isObject(target) && PrismaSelect.isObject(source)) {
      for (const key in source) {
        if (PrismaSelect.isObject(source[key])) {
          if (!target[key]) Object.assign(target, { [key]: {} });
          PrismaSelect.mergeDeep(target[key], source[key]);
        } else {
          Object.assign(target, { [key]: source[key] });
        }
      }
    }

    return PrismaSelect.mergeDeep(target, ...sources);
  }

  /**
   * Get nested value from select object.
   * @param field - name of field in select object.
   * @param filterBy - Model name as you have in schema.prisma file.
   * @param mergeObject
   * @example
   * // Graphql query
   * {
   *    findManyUser{
   *      id
   *      posts{
   *        id
   *        comments{
   *          id
   *        }
   *      }
   *    }
   * }
   *
   * // when you need to get more nested fields just add `.`
   * PrismaSelect.valueOf('posts.comments', 'Comment');
   * // return
   * { select: { id: true } }
   *
   * PrismaSelect.valueOf('posts', 'Post');
   *
   * // return
   * { select: { id: true, comments: { select: { id: true } } } }
   *
   **/
  valueOf(field: string, filterBy?: string, mergeObject: any = {}) {
    const splitItem = field.split('.');
    let newValue: Record<string, any> = this.getSelect(this.fields);
    for (const field of splitItem) {
      if (this.isAggregate && Object.prototype.hasOwnProperty.call(newValue, field)) {
        newValue = newValue[field];
      } else if (
        !this.isAggregate &&
        Object.prototype.hasOwnProperty.call(newValue, 'select') &&
        Object.prototype.hasOwnProperty.call(newValue.select, field)
      ) {
        newValue = newValue.select[field];
      } else {
        return {};
      }
    }
    return filterBy ? PrismaSelect.mergeDeep(this.filterBy(filterBy, newValue), mergeObject) : newValue;
  }

  /**
   * Work with this method if your GraphQL type name not like Schema model name.
   * @param modelName - Model name as you have in schema.prisma file.
   * @example
   * // normal call
   * const select = new PrismaSelect(info).value
   *
   * // With filter will filter select object with provided schema model name
   * const select = new PrismaSelect(info).valueWithFilter('User');
   *
   **/
  valueWithFilter(modelName: string) {
    return this.filterBy(modelName, this.getSelect(this.fields));
  }

  private filterBy(modelName: string, selectObject: any) {
    const model = this.model(modelName);
    if (model && typeof selectObject === 'object') {
      let defaultFields = {};
      if (this.defaultFields && this.defaultFields[modelName]) {
        const modelFields = this.defaultFields[modelName];
        defaultFields = typeof modelFields === 'function' ? modelFields(selectObject.select) : modelFields;
      }
      const filteredObject = {
        ...selectObject,
        select: { ...defaultFields },
      };
      Object.keys(selectObject.select).forEach((key) => {
        if (this.allowedProps.includes(key)) {
          filteredObject.select[key] = selectObject.select[key];
        } else {
          const field = this.field(key, model);
          if (field) {
            if (field.kind !== 'object') {
              filteredObject.select[key] = true;
            } else {
              const subModelFilter = this.filterBy(field.type, selectObject.select[key]);
              if (subModelFilter === true) {
                filteredObject.select[key] = true;
              } else if (Object.keys(subModelFilter.select).length > 0) {
                filteredObject.select[key] = subModelFilter;
              }
            }
          }
        }
      });
      return filteredObject;
    } else {
      return selectObject;
    }
  }

  private getArgs(args?: Record<string, unknown>) {
    const filteredArgs: Record<string, any> = {};
    if (args) {
      this.availableArgs.forEach((key) => {
        if (args[key]) {
          filteredArgs[key] = args[key];
        }
      });
    }
    return filteredArgs;
  }

  private getSelect(fields: PrismaSelect['fields']) {
    const selectObject: any = this.isAggregate ? {} : { select: {}, ...this.getArgs(fields?.args) };
    if (fields) {
      Object.keys(fields.fieldsByTypeName).forEach((type) => {
        const fieldsByTypeName = fields.fieldsByTypeName[type];
        Object.keys(fieldsByTypeName).forEach((key) => {
          const fieldName = fieldsByTypeName[key].name;
          if (Object.keys(fieldsByTypeName[key].fieldsByTypeName).length === 0) {
            if (this.isAggregate) {
              selectObject[fieldName] = true;
            } else {
              selectObject.select[fieldName] = true;
            }
          } else {
            if (this.isAggregate) {
              selectObject[fieldName] = this.getSelect(fieldsByTypeName[key]);
            } else {
              selectObject.select[fieldName] = this.getSelect(fieldsByTypeName[key]);
            }
          }
        });
      });
    }
    return selectObject;
  }
}
