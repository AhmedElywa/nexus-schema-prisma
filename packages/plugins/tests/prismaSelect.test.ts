import { server } from './server';
import gql from 'graphql-tag';
import { PrismaSelect } from '../src';
import { DocumentNode } from 'graphql';

const userQuery = gql`
  query user {
    user {
      id
      email
      name: firstName
      fullName
      posts(where: { title: { contains: "a" } }, orderBy: { published: asc }, take: 10, skip: 5, distinct: id) {
        id
        title
        comments(where: { contain: { contains: "a" } }) {
          id
          contain
        }
      }
    }
  }
`;

const mergeDeepQuery = gql`
  query mergeDeep {
    user {
      email
      posts {
        id
      }
    }
  }
`;

const mapQuery = gql`
  query valueOf {
    account {
      id
      firstName
      newFieldNotInSchema
    }
  }
`;

const aggregateUserQuery = gql`
  query aggregateUser {
    aggregateUser {
      _avg {
        id
      }
      _count {
        email
      }
      _max {
        id
      }
      _sum {
        id
      }
    }
  }
`;

const valueOfQuery = gql`
  query getNestedValue($type: String!, $value: String!) {
    getNestedValue(type: $type, value: $value) {
      id
      posts {
        id
        comments {
          id
          contain
        }
      }
    }
  }
`;

const userWithDefaultValuesQuery = gql`
  query userWithDefaultValues {
    userWithDefaultValues {
      id
    }
  }
`;

describe('test PrismaSelect class', () => {
  const executeOperation = async (query: DocumentNode, variables?: Record<string, any>) => {
    let log: { parsedResolveInfoFragment: any; select: any } = { parsedResolveInfoFragment: undefined, select: {} };
    const testServer = await server((o) => {
      log = o;
    });
    const result = await testServer.executeOperation({
      query,
      variables,
    });
    return {
      log,
      result,
    };
  };

  it('test nested relations with args', async () => {
    const { result, log } = await executeOperation(userQuery);
    expect(result.data).toMatchSnapshot();
    expect(result.errors).toMatchSnapshot();
    expect(log.select).toMatchSnapshot();
    expect(log.parsedResolveInfoFragment).toMatchSnapshot();
  });

  it('should merge deep add custom select object to current one', async () => {
    const { log } = await executeOperation(mergeDeepQuery);
    expect(
      PrismaSelect.mergeDeep(log.select, { select: { id: true, posts: { select: { title: true } } } }),
    ).toMatchSnapshot();
  });

  it('should get value from nested object and filter by given type', async () => {
    const { log } = await executeOperation(valueOfQuery, { value: 'posts', type: 'Post' });
    expect(log.select).toMatchSnapshot();
  });

  it('should get value from nested object more than one deep', async () => {
    const { log } = await executeOperation(valueOfQuery, { value: 'posts.comments', type: 'Comment' });
    expect(log.select).toMatchSnapshot();
  });

  it('should back empty object if the path not found', async () => {
    const { log } = await executeOperation(valueOfQuery, { value: 'post', type: 'Post' });
    expect(log.select).toMatchSnapshot();
  });

  it('should filter the type depend on the @PrismaSelect comment on the schema file', async () => {
    const { log } = await executeOperation(mapQuery);
    expect(log.select).toMatchSnapshot();
  });

  it('should add the selected fields on the args object if the type is aggregate', async () => {
    const { log } = await executeOperation(aggregateUserQuery);
    expect(log.select).toMatchSnapshot();
  });

  it('should contain default values firstName, lastName', async () => {
    const { log } = await executeOperation(userWithDefaultValuesQuery);
    expect(log.select).toMatchSnapshot();
  });
});
