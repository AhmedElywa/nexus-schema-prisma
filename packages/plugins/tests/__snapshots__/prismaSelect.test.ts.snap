// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`test PrismaSelect class should add the selected fields on the args object if the type is aggregate 1`] = `
Object {
  "_avg": Object {
    "id": true,
  },
  "_count": Object {
    "email": true,
  },
  "_max": Object {
    "id": true,
  },
  "_sum": Object {
    "id": true,
  },
}
`;

exports[`test PrismaSelect class should back empty object if the path not found 1`] = `Object {}`;

exports[`test PrismaSelect class should contain default values firstName, lastName 1`] = `
Object {
  "select": Object {
    "firstName": true,
    "id": true,
    "lastName": true,
  },
}
`;

exports[`test PrismaSelect class should filter the type depend on the @PrismaSelect comment on the schema file 1`] = `
Object {
  "select": Object {
    "firstName": true,
    "id": true,
  },
}
`;

exports[`test PrismaSelect class should get value from nested object and filter by given type 1`] = `
Object {
  "select": Object {
    "comments": Object {
      "select": Object {
        "contain": true,
        "id": true,
      },
    },
    "id": true,
  },
}
`;

exports[`test PrismaSelect class should get value from nested object more than one deep 1`] = `
Object {
  "select": Object {
    "contain": true,
    "id": true,
  },
}
`;

exports[`test PrismaSelect class should merge deep add custom select object to current one 1`] = `
Object {
  "select": Object {
    "email": true,
    "id": true,
    "posts": Object {
      "select": Object {
        "id": true,
        "title": true,
      },
    },
  },
}
`;

exports[`test PrismaSelect class test nested relations with args 1`] = `
Object {
  "user": null,
}
`;

exports[`test PrismaSelect class test nested relations with args 2`] = `undefined`;

exports[`test PrismaSelect class test nested relations with args 3`] = `
Object {
  "select": Object {
    "email": true,
    "firstName": true,
    "id": true,
    "posts": Object {
      "distinct": "id",
      "orderBy": Object {
        "published": "asc",
      },
      "select": Object {
        "comments": Object {
          "select": Object {
            "contain": true,
            "id": true,
          },
          "where": Object {
            "contain": Object {
              "contains": "a",
            },
          },
        },
        "id": true,
        "title": true,
      },
      "skip": 5,
      "take": 10,
      "where": Object {
        "title": Object {
          "contains": "a",
        },
      },
    },
  },
}
`;

exports[`test PrismaSelect class test nested relations with args 4`] = `
Object {
  "alias": "user",
  "args": Object {},
  "fieldsByTypeName": Object {
    "User": Object {
      "email": Object {
        "alias": "email",
        "args": Object {},
        "fieldsByTypeName": Object {},
        "name": "email",
      },
      "fullName": Object {
        "alias": "fullName",
        "args": Object {},
        "fieldsByTypeName": Object {},
        "name": "fullName",
      },
      "id": Object {
        "alias": "id",
        "args": Object {},
        "fieldsByTypeName": Object {},
        "name": "id",
      },
      "name": Object {
        "alias": "name",
        "args": Object {},
        "fieldsByTypeName": Object {},
        "name": "firstName",
      },
      "posts": Object {
        "alias": "posts",
        "args": Object {
          "distinct": "id",
          "orderBy": Object {
            "published": "asc",
          },
          "skip": 5,
          "take": 10,
          "where": Object {
            "title": Object {
              "contains": "a",
            },
          },
        },
        "fieldsByTypeName": Object {
          "Post": Object {
            "comments": Object {
              "alias": "comments",
              "args": Object {
                "where": Object {
                  "contain": Object {
                    "contains": "a",
                  },
                },
              },
              "fieldsByTypeName": Object {
                "Comment": Object {
                  "contain": Object {
                    "alias": "contain",
                    "args": Object {},
                    "fieldsByTypeName": Object {},
                    "name": "contain",
                  },
                  "id": Object {
                    "alias": "id",
                    "args": Object {},
                    "fieldsByTypeName": Object {},
                    "name": "id",
                  },
                },
              },
              "name": "comments",
            },
            "id": Object {
              "alias": "id",
              "args": Object {},
              "fieldsByTypeName": Object {},
              "name": "id",
            },
            "title": Object {
              "alias": "title",
              "args": Object {},
              "fieldsByTypeName": Object {},
              "name": "title",
            },
          },
        },
        "name": "posts",
      },
    },
  },
  "name": "user",
}
`;
